(serve.llm) Remove test leakage from placement bundle logic

nrghosh opened the pull request:
## Fixing Test Code Leakage in LLM Placement Bundle Logic

### Problem

**Test Code Leakage**: Production code had test-specific logic that didn't follow clean separation of concerns:
- Tests specified `accelerator_type: A10G` in YAML configs, causing `placement_bundles` to return `[{"GPU": 1, "accelerator_type:A10G": 0.001}]`
- In CI environments without actual GPU hardware, Ray couldn't schedule these resources
- Workaround `mock_resource: 0` check was embedded in production code to bypass this and set `bundles = []`
- Test infrastructure was coupled to production logic, which is not ideal

### Solution

**Clean Test/Production Separation**: Implemented proper test fixtures that isolate test behavior without modifying production code:

1. **Test Fixture Approach** (`conftest.py`):
   - Created `disable_placement_bundles` fixture using `unittest.mock.patch`
   - Patches `VLLMEngineConfig.placement_bundles` property to return `[]` during tests
   - Automatically applied to test fixtures that need GPU-free execution

2. **Subprocess Test Handling**:
   - Initial monkey-patch only worked for in-process pytest tests
   - Added explicit CPU resource configuration for subprocess tests (`serve run`)
   - Used `resources_per_bundle: {"CPU": 1}` for CPU-only testing scenarios

3. **Production Code Cleanup**:
   - Removed `mock_resource` checks from `server_models.py`
   - Eliminated test-specific environment variables that leaked into production
   - Restored clean placement bundle logic for production use

### Test Coverage

**Fixed Test Categories**:
- CPU tests using mock engines (no longer hang waiting for GPU resources)
- Subprocess tests using `serve run` (proper resource declarations)
- GPU integration tests with mock engines (clean separation of mock vs real engines)

**Key Tests Fixed**:
- `test_build_openai_app_with_config` (CPU subprocess test)
- `test_lora_unavailable_base_model` (CPU mock test) 
- `test_models` (GPU integration test with mock engine)

### Architecture

**Before (Test Leakage)**:
```python
# Production code contained test logic
if "mock_resource" in config:
    return []  # Test-specific behavior in prod!
```

**After (Clean Separation)**:
```python
# Test fixture handles test-specific behavior
@pytest.fixture
def disable_placement_bundles():
    with patch.object(VLLMEngineConfig, "placement_bundles", 
                     new_callable=lambda: property(lambda self: [])):
        yield
```

### Approach Evolution

**Earlier broad monkey-patching broke existing tests**, which led to a more refined approach:
- Use targeted resource declarations (`resources_per_bundle: {"CPU": 1}`) for CPU-only scenarios
- Preserve all core placement bundle functionality for production
- Keep backward compatibility

nrghosh committed:
Remove test code leakage from placement group logic

- Replace hacky fix to bypass placement groups when mock_resource was
  present
- Remove coupling between test infra and prod code
- Remove mock_resource checks, add test fixture to configure
  placement_bundles as [] when in testing
- Tests now run without requiring GPUs by using a test-layer patch
  instead of modifying production logic
- Addresses TODO (server_models.py)

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh commented:
**Problem:**
Looking at the placement_bundles implementation -> can see the following issue
- Tests specify accelerator_type: A10G in YAML configs
- Causes engine_config.placement_bundles to return bundles like [{"GPU": 1, "accelerator_type:A10G": 0.001}]
- In CI environments without actual A10G hardware, Ray can't schedule these resources
- The mock_resource: 0 hack was used to bypass this and set bundles = []

**Fix:**
- Monkey-patch the placement_bundles property so it returns empty bundles (so we don't have to make changes to prod code)
- Test Fixture (conftest.py): Creates an autouse=True pytest fixture that automatically patches the VLLMEngineConfig.placement_bundles property so that the patch returns empty bundles ([]) for all test scenarios
- Clean up prod code / hacky YAML file fixes / test files

**Testing Passed:**
- [x] `python/ray/llm/tests/serve/cpu/builders/test_application_builders.py`
- [x] `python/ray/llm/tests/serve/cpu/deployments/llm/multiplex/test_lora_deployment_base_client.py`
- [x] `python/ray/llm/tests/serve/gpu/integration/test_openai_compatibility.py`

nrghosh committed:
Fix CI Testing Failure: Pytest vs Serve Run

- Root Cause: Patch in conftest.py only worked within the pytest process, but the failing test (test_build_openai_app_with_config) spawned a subprocess using serve run, which didn't inherit the monkey patch
- Monkey patch for in-process tests: Continues to work for regular pytest tests
- Environment variable for subprocesses: RAYLLM_TEST_DISABLE_PLACEMENT_GROUPS=1 is set by the fixture and checked
by the production code

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

eicherseiji commented on python/ray/llm/_internal/serve/deployments/llm/vllm/vllm_models.py:
```diff
@@ -142,6 +142,12 @@ def placement_strategy(self) -> str:
+        import os
+
+        # Check if we're in a test environment that wants to disable placement groups
+        if os.environ.get("RAYLLM_TEST_DISABLE_PLACEMENT_GROUPS") == "1":
```
Do we still need this/can we avoid creating `RAYLLM_TEST_DISABLE_PLACEMENT_GROUPS` completely using the mock/monkeypatch method? I feel this is still a bit test-leaky.

nrghosh replied:
yep, working through fixes

nrghosh committed:
address test / env-var leakage

- stop using env variable to control placement groups
- monkey patching (global) broke tests by overriding all
- tests ensure placement bundle logic works
- eliminate leakage while still maintaining test coverage
- fix tests by relaxing some assertions on serve_options dict

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
working update for tests + fix timeout

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
fix breaking test

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

kouroshHakha reviewed (approved):
LGTM. waiting for the tests.

nrghosh committed:
ensure serve run subprocs and in-process api calls both mock out gpu resource needs. apply placement bundle override to overlooked tests, fix type errors

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
fix: Add disable_placement_bundles to mock engine tests

The refactoring to explicit placement group mocking missed some fixtures
that use mock engines but still tried to create real placement groups.

Changes:
- Add disable_placement_bundles dependency to testing_model fixtures in conftest.py
- Set accelerator_type: None in test_application_builders.py subprocess configs
- Remove environment variable approach to avoid test code leakage

This fixes hanging tests in both CPU and GPU test suites that use mock
engines (MockVLLMEngine) but were attempting to create real GPU placement
groups. Mock vs real engine tests need different placement
 group handling, not CPU vs GPU tests.

Fixes:
- test_build_openai_app_with_config (CPU subprocess test)
- test_lora_unavailable_base_model (CPU mock test)
- test_models (GPU integration test with mock engine)

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
fix: Use explicit CPU resources in test configs to avoid GPU placement groups

- fix broad initial approach that was breaking some tests
- more targeted approach (explicitly declare test resources in config)
- Revert core placement_bundles logic to preserve production behavior
- Add resources_per_bundle: {"CPU": 1} to test fixture for CPU-only testing
- Fixes test_build_openai_app_with_config deployment timeouts without breaking existing tests

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge branch 'master' into fix-test_code_leakage

Signed-off-by: Nikhil G <nrghosh@users.noreply.github.com>

nrghosh committed:
lint

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge branch 'master' into fix-test_code_leakage
